internal/helm/chart/builder_local.go

```diff
@@ -163,7 +164,7 @@ func (b *localChartBuilder) Build(ctx context.Context, ref Reference, p string,
 		if err != nil {
 			return result, &BuildError{Reason: ErrValuesFilesMerge, Err: err}
 		}
-		result.ValuesFiles = opts.GetValuesFiles()
+		result.ValuesFiles = valuesFiles
```

Comment:
Thanks for adapting this.

Unfortunately there is still an issue. When `if opts.CachedChart != "" && !opts.Force` holds true, there can be an early return with the cached chart. When this happens, we have again the wrong summary. The scenario where this matters is when recovering from a failure from the last reconciliation, if we have a cached chart (we don't repackage) with missing files, the summary will be wrong. To fix that we need to persist `mergeFileValues` result and use it when returning cached chart result.

Reply:
How and where do you suggest we persist mergeFileValues?

Reply:
@pcanilho has [tried persisting via `BuildOptions`](https://github.com/isometry/source-controller/pull/1/files), but as the `Builder` interface's `Build` method passes `BuildOptions` by copy rather than reference, we don't think this is sufficient as local mutations will still be lost between invocations.

@souleb: can you propose a better approach? A breaking change to the `Builder` interface certainly seems "suboptimal".

Reply:
If there are missing `valuesFiles`, we should persist the used `valuesFile` list in order to be able to retrieve it between reconciliation. Usually this is done in `obj.Status`. I think this would need to happen in `reconcileArtifact` here: https://github.com/fluxcd/source-controller/blob/ec85f3af4c679f8c1de3b5de1f4d488327000686/internal/controller/helmchart_controller.go#L886.

In the next reconciliation, we would pass the persisted list, if it exists as part of the `BuildOptions` (otherwise keep current behavior).

Reply:
Great. If you're OK with the addition of an `ObservedValuesFiles` field on the `HelmChartStatus` struct to act as the persistent store, I'll work on that.

Reply:
I have started working on the needed changes here: https://github.com/isometry/source-controller/pull/2/files
I will be doing a pass over the existing tests to see where/what needs adjusting.

Reply:
Great. Don't hesitate to ping me if needed.

Reply:
@souleb … hopefully @pcanilho's patch addresses this.